### PR TITLE
Refactors defender variable handling

### PR DESCRIPTION
**Pull request by carlarctg**
# About the pull request

Refactors defender variable handling, turning variables that are on Xenomorph into behaviour delegate variables.

Deletes some stupid poop vehicle code that apparently was supposed to make fortified defenders not stop the APC but doesn't actually do that.

# Explain why it's good for the game

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog

:cl:
refactor: Refactors defender variable handling, turning variables that are on Xenomorph into behaviour delegate variables.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/Evolution.dm [hunk 1]
```diff
@@ -197,9 +197,8 @@
 		to_chat(src, SPAN_WARNING("You must be at full health to evolve."))
 		return FALSE
 
-	if(agility || fortify || crest_defense)
-		to_chat(src, SPAN_WARNING("You cannot evolve while in this stance."))
-		return FALSE
+	if(!behavior_delegate.on_evolve())
+		return
```
Comment (mullenpaul) [inline code/modules/mob/living/carbon/xenomorph/Evolution.dm:201]: Shouldn't this return FALSE to remain consistent with the rest of  the proc?

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 2]
```diff
@@ -13,17 +14,17 @@
 	if(!action_cooldown_check())
 		return
 
-	xeno.crest_defense = !xeno.crest_defense
-
-	if(xeno.crest_defense)
+	if(defender_delegate.crest_lowered == FALSE)
```
Comment (mullenpaul) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:17]: Don't need to explicitly check for FALSE

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 3]
```diff
@@ -1,9 +1,10 @@
 /datum/action/xeno_action/onclick/toggle_crest/use_ability(atom/target)
 	var/mob/living/carbon/Xenomorph/xeno = owner
+	var/datum/behavior_delegate/defender_base/defender_delegate = xeno.behavior_delegate
```
Comment (mullenpaul) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:3]: Something feels off - I don't think it makes much sense to refactor the defender specific variables into a delegate, only to hard code that delegate to the defender. It's nice to pull vars out of the defender that don't need to be there but this makes me think it's an exercise in refactoring for the sake of it

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 4]
```diff
@@ -62,12 +64,14 @@
...
-	var/max_distance = 3 - (fendy.crest_defense * 2)
+	var/max_distance = 3
+	if(defender_delegate.crest_lowered)
+		max_distance = 1
 
 	if(distance > max_distance)
 		return
 
-	if(!fendy.crest_defense)
+	if(defender_delegate.crest_lowered == FALSE)
```
Comment (mullenpaul) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:74]: Again, don't need to explicitly check for FALSE

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 5]
```diff
@@ -78,12 +82,18 @@
...
 	if(carbone.stat != DEAD && (!(carbone.status_flags & XENO_HOST) || !HAS_TRAIT(carbone, TRAIT_NESTED)) )
-		var/h_damage = 30 - (fendy.crest_defense * 10) + (fendy.steelcrest * 7.5) //30 if crest up, 20 if down, plus 7.5
+		var/h_damage = fendy.caste.melee_damage_upper + crest_damage + strain_damage
 		carbone.apply_armoured_damage(get_xeno_damage_slash(carbone, h_damage), ARMOR_MELEE, BRUTE, "chest", 5)
 
 	var/facing = get_dir(fendy, carbone)
-	var/headbutt_distance = 1 + (fendy.crest_defense * 2) + (fendy.fortify * 2)
+	var/headbutt_distance = 1
+	if(defender_delegate.crest_lowered || defender_delegate.fortified)
+		headbutt_distance += 2
```
Comment (mullenpaul) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:96]: Why not explicitly state 3?

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 6]
```diff
@@ -105,41 +115,42 @@
 
 // Defender Tail Sweep
 /datum/action/xeno_action/onclick/tail_sweep/use_ability(atom/A)
-	var/mob/living/carbon/Xenomorph/X = owner
-	if (!istype(X))
+	var/mob/living/carbon/Xenomorph/fender = owner
+	var/datum/behavior_delegate/defender_base/defender_delegate = fender.behavior_delegate
```
Comment (mullenpaul) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:119]: Again, I think the delegate abstraction is off

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 7]
```diff
@@ -175,7 +183,7 @@
 
 	playsound(get_turf(xeno), 'sound/effects/stonedoor_openclose.ogg', 30, 1)
 
-	if(!xeno.fortify)
+	if(defender_delegate.fortified == FALSE)
```
Comment (mullenpaul) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:186]: Don't need to explicitly check FALSE

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm [hunk 8]
```diff
@@ -193,63 +201,67 @@
...
 	if(fortify_state)
-		to_chat(X, SPAN_XENOWARNING("You tuck yourself into a defensive stance."))
-		if(X.steelcrest)
-			X.armor_deflection_buff += 10
-			X.armor_explosive_buff += 60
-			X.ability_speed_modifier += 3
-			X.damage_modifier -= XENO_DAMAGE_MOD_SMALL
+		to_chat(fender, SPAN_XENOWARNING("You tuck yourself into a defensive stance."))
+		if(defender_delegate.steelcrest_strain)
+			fender.armor_deflection_buff += 10
```
Comment (mullenpaul) [inline code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:223]: I think these numbers should be refactored into datum vars - they are hidden in the code and the numbers are duplicated.

**Review by mullenpaul — CHANGES_REQUESTED** [review]
I think there is an issue with the abstraction, you are explicitly coupling the xeno_action to the behavior_delegate subclass which removes the flexibility of injecting behaviour with the delegate. This means that these abilities could only ever work with the defender

File: code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm (hunk 2 above)
Comment (carlarctg) [reply code/modules/mob/living/carbon/xenomorph/abilities/defender/defender_powers.dm:17]: It's for clarity

**Pull request closed without merging**